dask/highlevelgraph.py

```diff
@@ -24,6 +27,122 @@ def _find_layer_containing_key(key):
     return ret
 
 
+class Layer(Mapping):
+    """High level graph layer
+
+    This abstract class establish a protocol for high level graph layers.
+    """
+
+    def cull(self, keys: Set) -> "Layer":
+        """Return a new Layer with only the tasks required to calculate `keys`.
+
+        In other words, remove unnecessary tasks from the layer.
+
+        Examples
+        --------
+        >>> d = Layer({'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)})  # doctest: +SKIP
+        >>> d.cull({'out'})  # doctest: +SKIP
+        {'x': 1, 'out': (add, 'x', 10)}
+
+        Returns
+        -------
+        layer: Layer
+            Culled layer
+        """
+
+        seen = set()
+        out = {}
+        work = set(flatten(keys))
+        while work:
+            new_work = set()
+            for k in work:
+                out[k] = self[k]
+                for d in get_dependencies(self, k):
+                    if d not in seen:
+                        seen.add(d)
+                        new_work.add(d)
+            work = new_work
+
+        return BasicLayer(out)
+
+    def get_external_dependencies(self, all_hlg_keys: Set) -> Set:
+        """Get external dependencies
+
+        Parameters
+        ----------
+        all_hlg_keys : set
+            Set of all keys in the high level graph.
+
+        Returns
+        -------
+        deps: set
+            Set of dependencies
+        """
+
+        all_deps = keys_in_tasks(all_hlg_keys, self.values())
+        return all_deps.difference(self.keys())
```

Comment:
Maybe this is an optimization for the future, but could we restrict ourselves to the dependencies in the high level layer dependencies?

Reply:
Let's postpone this to the future. 
My guess is that the run time is dominated by the size of `self.values()` but I don't see an obvious way to reduce it before calling `keys_in_tasks()`?